src/main/java/com/google/sps/utility/LocationsUtility.java

```diff
@@ -63,4 +64,39 @@ private static String getLocation(Pattern pattern, String prefix, String taskNot
     }
     return "";
   }
+
+  /**
+   * Generates combinations of the lists and appends them to the result parameter.
+   *
+   * @param lists A list of lists to generate combinations for.
+   * @param result A pointer to a list to contain the generated combinations.
+   * @param depth An int to represent the index of the list that is currently being traversed in the
+   *     recursive call.
+   * @param current A list containing the combination built so far at the recursive call.
+   */
+  private static void generateCombinationsHelper(
+      List<List<String>> lists, List<List<String>> result, int depth, List<String> current) {
+    if (depth == lists.size()) {
+      result.add(new ArrayList<>(current));
+      return;
+    }
+    for (int i = 0; i < lists.get(depth).size(); i++) {
+      List<String> next = new ArrayList<>();
+      next.addAll(current);
+      next.add(lists.get(depth).get(i));
+      generateCombinationsHelper(lists, result, depth + 1, next);
```

Comment:
Currently, you do O(|current|^2) work for each `current` list you end up adding to result because all the elements get copied over from current to next every time a new element gets added.
You can drop that to O(|current|) if you just append the new element to current and then delete it. I.e.,
```
current.add(lists.get(depth).get(i));
generateCombinationsHelper(lists, result, depth + 1, current);
current.remove(depth);
```
Let me know if you have any questions about this trick.

Reply:
This is a smart trick, I assume it's better in terms of space complexity as well. Thanks!

Reply:
`RESOLVED`